### PR TITLE
[minor] fix multi domain fetch

### PR DESCRIPTION
**Pull request by WindzCUHK**
## TODO

- [x] unit test
- [x] add comments

## Design
`fetcher.go`
1. handle HTTP, etag related logic
2. `policyCache` as cache for etag and domain policy

`daemon.go`
1. remove `etagCache`, `etagFlushDur`
1. add `fetchers`, each handle a single domain

[file shortcut](https://github.com/yahoojapan/athenz-authorizer/pull/38/files?file-filters%5B%5D=.go)

### CODE REVIEW

**Comment by WindzCUHK** [comment]
will fix the fetch on start logic in https://github.com/yahoojapan/athenz-authorizer/pull/46 later

**Comment by WindzCUHK** [comment]
# sample log msg in proxy currently
should also print the whole map in debug log after the `tmp cache becomes effective` msg?
```
2019-08-22 14:41:44	[INFO]:	Update ZTS athenz pubkey success
2019-08-22 14:41:44	[INFO]:	Starting policyd updater
2019-08-22 14:41:44	[INFO]:	[1566452504] will update policy
2019-08-22 14:41:44	[INFO]:	will fetch policy for domain: yby.wfan.play
2019-08-22 14:41:44	[INFO]:	will fetch policy for domain: yby.wfan.shopping
2019-08-22 14:41:44	[DEBG]:	will fetch policy from url: https://zts.athens.yahoo.co.jp:4443/zts/v1/domain/yby.wfan.shopping/signed_policy_data
2019-08-22 14:41:44	[DEBG]:	will fetch policy from url: https://zts.athens.yahoo.co.jp:4443/zts/v1/domain/yby.wfan.play/signed_policy_data
2019-08-22 14:41:44	[DEBG]:	set policy cache for domain: yby.wfan.play, policy: { ctime: 2019-08-22 14:41:44.672382 +0900 JST, etag: "2019-07-26T08:21:43.580Z", etagExpiry: 2019-08-27 05:41:45.456 +0000 UTC, sp.domain: yby.wfan.play }
2019-08-22 14:41:44	[DEBG]:	set policy cache for domain: yby.wfan.shopping, policy: { ctime: 2019-08-22 14:41:44.672382 +0900 JST, etag: "2018-06-06T02:40:14.719Z", etagExpiry: 2019-08-27 05:41:45.457 +0000 UTC, sp.domain: yby.wfan.shopping }
2019-08-22 14:41:44	[DEBG]:	will merge policy, domain: yby.wfan.play, body: {"keyId":"0","signature":"...","signedPolicyData":{"expires":"2019-08-29T05:41:45.456Z","modified":"2019-07-26T08:21:43.580Z","policyData":{"domain":"yby.wfan.play","policies":[{"assertions":[{"action":"*","effect":"ALLOW","resource":"yby.wfan.play:*","role":"yby.wfan.play:role.admin"}],"modified":"2018-07-18T03:02:01.274Z","name":"yby.wfan.play:policy.admin"},{"assertions":[{"action":"*","effect":"ALLOW","resource":"yby.wfan.play:*","role":"yby.wfan.play:role.zcp-tester"},{"action":"get","effect":"ALLOW","resource":"yby.wfan.play:*","role":"yby.wfan.play:role.zcp-tester"},{"action":"get","effect":"ALLOW","resource":"yby.wfan.play:rrr","role":"yby.wfan.play:role.zcp-tester"}],"modified":"2019-07-26T08:21:43.578Z","name":"yby.wfan.play:policy.zcp-tester"}]},"zmsKeyId":"0","zmsSignature":"..."}}
2019-08-22 14:41:44	[DEBG]:	will merge policy, domain: yby.wfan.shopping, body: {"keyId":"0","signature":"...","signedPolicyData":{"expires":"2019-08-29T05:41:45.457Z","modified":"2018-06-06T02:40:14.719Z","policyData":{"domain":"yby.wfan.shopping","policies":[{"assertions":[{"action":"get","effect":"ALLOW","resource":"yby.wfan.shopping:result.v1.name","role":"yby.wfan.shopping:role.users"}],"modified":"2018-06-06T02:37:54.973Z","name":"yby.wfan.shopping:policy.item-name"},{"assertions":[{"action":"*","effect":"ALLOW","resource":"yby.wfan.shopping:*","role":"yby.wfan.shopping:role.admin"}],"modified":"2018-06-06T02:36:21.661Z","name":"yby.wfan.shopping:policy.admin"}]},"zmsKeyId":"0","zmsSignature":"..."}}
2019-08-22 14:41:44	[DEBG]:	added assertion to the tmp cache: &{Action:* Effect:ALLOW Id:0 Resource:yby.wfan.play:* Role:yby.wfan.play:role.zcp-tester}
2019-08-22 14:41:44	[DEBG]:	added assertion to the tmp cache: &{Action:get Effect:ALLOW Id:0 Resource:yby.wfan.play:* Role:yby.wfan.play:role.zcp-tester}
2019-08-22 14:41:44	[DEBG]:	added assertion to the tmp cache: &{Action:* Effect:ALLOW Id:0 Resource:yby.wfan.play:* Role:yby.wfan.play:role.admin}
2019-08-22 14:41:44	[DEBG]:	added assertion to the tmp cache: &{Action:* Effect:ALLOW Id:0 Resource:yby.wfan.shopping:* Role:yby.wfan.shopping:role.admin}
2019-08-22 14:41:44	[DEBG]:	added assertion to the tmp cache: &{Action:get Effect:ALLOW Id:0 Resource:yby.wfan.play:rrr Role:yby.wfan.play:role.zcp-tester}
2019-08-22 14:41:44	[DEBG]:	added assertion to the tmp cache: &{Action:get Effect:ALLOW Id:0 Resource:yby.wfan.shopping:result.v1.name Role:yby.wfan.shopping:role.users}
2019-08-22 14:41:44	[DEBG]:	tmp cache becomes effective
2019-08-22 14:41:44	[INFO]:	[1566452504] update policy done
2019-08-22 14:41:54	[INFO]:	[1566452514] will update policy
2019-08-22 14:41:54	[INFO]:	will fetch policy for domain: yby.wfan.shopping
2019-08-22 14:41:54	[INFO]:	will fetch policy for domain: yby.wfan.play
2019-08-22 14:41:54	[DEBG]:	will fetch policy from url: https://zts.athens.yahoo.co.jp:4443/zts/v1/domain/yby.wfan.play/signed_policy_data
2019-08-22 14:41:54	[DEBG]:	request on domain: yby.wfan.play, with etag: "2019-07-26T08:21:43.580Z"
2019-08-22 14:41:54	[DEBG]:	will fetch policy from url: https://zts.athens.yahoo.co.jp:4443/zts/v1/domain/yby.wfan.shopping/signed_policy_data
2019-08-22 14:41:54	[DEBG]:	request on domain: yby.wfan.shopping, with etag: "2018-06-06T02:40:14.719Z"
2019-08-22 14:41:54	[DEBG]:	policy = 304 not modified, use cache for domain: yby.wfan.play, etag: "2019-07-26T08:21:43.580Z"
2019-08-22 14:41:54	[DEBG]:	policy = 304 not modified, use cache for domain: yby.wfan.shopping, etag: "2018-06-06T02:40:14.719Z"
2019-08-22 14:41:54	[DEBG]:	will merge policy, domain: yby.wfan.play, body: {"keyId":"0","signature":"...","signedPolicyData":{"expires":"2019-08-29T05:41:45.456Z","modified":"2019-07-26T08:21:43.580Z","policyData":{"domain":"yby.wfan.play","policies":[{"assertions":[{"action":"*","effect":"ALLOW","resource":"yby.wfan.play:*","role":"yby.wfan.play:role.admin"}],"modified":"2018-07-18T03:02:01.274Z","name":"yby.wfan.play:policy.admin"},{"assertions":[{"action":"*","effect":"ALLOW","resource":"yby.wfan.play:*","role":"yby.wfan.play:role.zcp-tester"},{"action":"get","effect":"ALLOW","resource":"yby.wfan.play:*","role":"yby.wfan.play:role.zcp-tester"},{"action":"get","effect":"ALLOW","resource":"yby.wfan.play:rrr","role":"yby.wfan.play:role.zcp-tester"}],"modified":"2019-07-26T08:21:43.578Z","name":"yby.wfan.play:policy.zcp-tester"}]},"zmsKeyId":"0","zmsSignature":"..."}}
2019-08-22 14:41:54	[DEBG]:	will merge policy, domain: yby.wfan.shopping, body: {"keyId":"0","signature":"...","signedPolicyData":{"expires":"2019-08-29T05:41:45.457Z","modified":"2018-06-06T02:40:14.719Z","policyData":{"domain":"yby.wfan.shopping","policies":[{"assertions":[{"action":"get","effect":"ALLOW","resource":"yby.wfan.shopping:result.v1.name","role":"yby.wfan.shopping:role.users"}],"modified":"2018-06-06T02:37:54.973Z","name":"yby.wfan.shopping:policy.item-name"},{"assertions":[{"action":"*","effect":"ALLOW","resource":"yby.wfan.shopping:*","role":"yby.wfan.shopping:role.admin"}],"modified":"2018-06-06T02:36:21.661Z","name":"yby.wfan.shopping:policy.admin"}]},"zmsKeyId":"0","zmsSignature":"..."}}
2019-08-22 14:41:54	[DEBG]:	added assertion to the tmp cache: &{Action:* Effect:ALLOW Id:0 Resource:yby.wfan.play:* Role:yby.wfan.play:role.zcp-tester}
2019-08-22 14:41:54	[DEBG]:	added assertion to the tmp cache: &{Action:get Effect:ALLOW Id:0 Resource:yby.wfan.play:* Role:yby.wfan.play:role.zcp-tester}
2019-08-22 14:41:54	[DEBG]:	added assertion to the tmp cache: &{Action:get Effect:ALLOW Id:0 Resource:yby.wfan.shopping:result.v1.name Role:yby.wfan.shopping:role.users}
2019-08-22 14:41:54	[DEBG]:	added assertion to the tmp cache: &{Action:get Effect:ALLOW Id:0 Resource:yby.wfan.play:rrr Role:yby.wfan.play:role.zcp-tester}
2019-08-22 14:41:54	[DEBG]:	added assertion to the tmp cache: &{Action:* Effect:ALLOW Id:0 Resource:yby.wfan.play:* Role:yby.wfan.play:role.admin}
2019-08-22 14:41:54	[DEBG]:	added assertion to the tmp cache: &{Action:* Effect:ALLOW Id:0 Resource:yby.wfan.shopping:* Role:yby.wfan.shopping:role.admin}
2019-08-22 14:41:54	[DEBG]:	tmp cache becomes effective
2019-08-22 14:41:54	[INFO]:	[1566452514] update policy done
```

File: model.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ limitations under the License.
```
Comment (kevindiu) [inline model.go]: maybe we can remove this file :)

File: policy/daemon.go [hunk 2]
```diff
@@ -141,25 +146,26 @@ func (p *policyd) Start(ctx context.Context) <-chan error {
 
 // Update updates and cache policy data
 func (p *policyd) Update(ctx context.Context) error {
-	glg.Info("Updating policy")
-	defer glg.Info("Updated policy")
+	jobID := fastime.Now().Unix()
+	glg.Infof("[%d] will update policy", jobID)
+	defer glg.Infof("[%d] update policy done", jobID)
```
Comment (kevindiu) [inline policy/daemon.go]: suggest move it to the bottom of the function instead of using defer.

**Comment by kevindiu** [comment]
TODO: remove first blocking Update() in Start() function

File: policy/daemon.go [hunk 3]
```diff
@@ -141,45 +146,48 @@ func (p *policyd) Start(ctx context.Context) <-chan error {
 
 // Update updates and cache policy data
 func (p *policyd) Update(ctx context.Context) error {
-	glg.Info("Updating policy")
-	defer glg.Info("Updated policy")
+	jobID := fastime.Now().Unix()
```
Comment (kevindiu) [inline policy/daemon.go]: can we delete this jobID?

File: policy/daemon.go [hunk 4]
```diff
@@ -141,45 +146,48 @@ func (p *policyd) Start(ctx context.Context) <-chan error {
...
 				default:
-					return p.fetchAndCachePolicy(ctx, rp, dom)
+					return fetchAndCachePolicy(ctx, rp, f)
 				}
 			})
 		}
 	}
 
 	if err := eg.Wait(); err != nil {
+		glg.Warnf("[%d] update policy fail", jobID)
```
Comment (kevindiu) [inline policy/daemon.go]: if shd be glg.Errorf() instead of Warnf

File: policy/daemon.go (hunk 3 above)
Comment (WindzCUHK) [reply policy/daemon.go]: It is easier to track the start and end from the log.
Better to used UUID other than timestamp, just that timestamp is simpler this time.

File: policy/daemon.go (hunk 4 above)
Comment (WindzCUHK) [reply policy/daemon.go]: will fix

Comment (WindzCUHK) [reply policy/daemon.go]: fixed

File: policy/fetcher.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	retryInterval time.Duration
+	retryMaxCount int
+
+	// athenz related
+	domain     string
+	athenzURL  string
+	spVerifier SignedPolicyVerifier
+
+	client      *http.Client
+	policyCache unsafe.Pointer
```
Comment (kevindiu) [inline policy/fetcher.go]: we should use *atomic.Value instead of unsafe.Pointer here

File: policy/daemon.go [hunk 6]
```diff
@@ -85,6 +75,23 @@ func New(opts ...Option) (Daemon, error) {
...
 	}
 
+	// create fetchers
+	p.fetchers = make(map[string]Fetcher, len(p.athenzDomains))
+	for _, domain := range p.athenzDomains {
+		f := fetcher{
+			domain:        domain,
+			expireMargin:  p.expireMargin,
+			retryInterval: p.errRetryInterval,
+			retryMaxCount: 3,
```
Comment (kevindiu) [inline policy/daemon.go]: should we make it configurable?

File: policy/fetcher.go (hunk 5 above)
Comment (WindzCUHK) [reply policy/fetcher.go]: they are the same?

unsafe seems faster. (only load is tested...)
https://blog.dubbelboer.com/2015/08/23/rwmutex-vs-atomicvalue-vs-unsafepointer.html

File: policy/daemon.go (hunk 6 above)
Comment (WindzCUHK) [reply policy/daemon.go]: should be.
This PR? It is an interface change.

Comment (kevindiu) [reply policy/daemon.go]: can be another PR.

Comment (WindzCUHK) [reply policy/daemon.go]: continue in https://github.com/yahoojapan/athenz-authorizer/pull/50

File: policy/daemon.go [hunk 7]
```diff
@@ -256,101 +260,30 @@ func (p *policyd) GetPolicyCache(ctx context.Context) map[string]interface{} {
...
 
-func (p *policyd) fetchAndCachePolicy(ctx context.Context, g gache.Gache, dom string) error {
-	spd, upd, err := p.fetchPolicy(ctx, dom)
+func fetchAndCachePolicy(ctx context.Context, g gache.Gache, f Fetcher) error {
+	sp, err := f.FetchWithRetry(ctx)
 	if err != nil {
-		glg.Debugf("fetch policy failed, err: %v", err)
-		return errors.Wrap(err, "error fetch policy")
+		errMsg := "fetch policy fail"
+		glg.Debugf("%s, error: %v", errMsg, err)
```
[the remark this answers is not in the document]
Comment (WindzCUHK) [reply policy/daemon.go]: `Warnf()`?

[the remark this answers is not in the document]
Comment (kevindiu) [reply policy/daemon.go]: arrr maybe errorf is best

Comment (kevindiu) [inline policy/daemon.go]:
```suggestion
		glg.Errorf("%s, error: %v", errMsg, err)
```

**Comment by kevindiu** [comment]
It is minor changes instead of patch?

**Comment by kevindiu** [comment]
otherwise it is LGTM and can be merged

**Comment by WindzCUHK** [comment]
@kevindiu updated PR title

**Pull request merged**